### PR TITLE
[286] Add zendesk integration

### PR DESCRIPTION
**Pull request by aliuk2012**
### Context
Integrate Zendesk into the service so that we are able to log zendesk tickets through our service instead of users having to send us emails.

Before merging this PR, two new environment vars for zendesk username & token will be added to the production instance for API calls to work. We plan on testing get support forms and this integration before soft-launching it across the service.

### CODE REVIEW

File: spec/services/zendesk_service_spec.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+  xdescribe '.send!' do
+    it 'calls #send!' do
+      # expect(described_class).to have_received(:new).with(ticket).and_return(service)
+      # expect(service).to have_received(:send!)
+      # described_class.send!(ticket)
+    end
+  end
+
+  xdescribe '#send!' do
```
Comment (aliuk2012) [inline spec/services/zendesk_service_spec.rb:30]: Could really do with some help with these tests

File: app/form_objects/support_ticket/check_your_request_form.rb [hunk 2]
```diff
@@ -6,7 +6,17 @@ class SupportTicket::CheckYourRequestForm
   validates :ticket, presence: true
 
   def create_ticket
-    ticket['subject'] = "TESTING - (#{ticket['school_unique_id']}) #{ticket['school_name']} "
-    ZendeskService.send!(ticket)
+    if Settings.zendesk.present? && Settings.zendesk.username.present? && Settings.zendesk.token.present?
```
Comment (aldavidson) [inline app/form_objects/support_ticket/check_your_request_form.rb:9]:
```suggestion
    if Settings.zendesk&.username.present? && Settings.zendesk&.token.present?
```

Comment (aldavidson) [reply app/form_objects/support_ticket/check_your_request_form.rb:9]: Just a bit more concise

File: app/services/zendesk_service.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
+class ZendeskService
+  attr_accessor :ticket
+
+  class << self
+    def send!(ticket)
+      ZendeskService.new(ticket).send!
```
Comment (aldavidson) [inline app/services/zendesk_service.rb:6]:
```suggestion
      new(ticket).send!
```
It's already in the ZendeckService class scope, so you don't need to re-reference it

File: app/services/zendesk_service.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      requester: { email: ticket['email_address'], name: ticket['full_name'] },
+      subject: ticket['subject'],
+      comment: {
+        body: ticket['message'],
+      },
+      custom_fields: [
+        { id: '360011490478', value: 'contact_form' },
+        { id: '360011798678', value: ticket['user_type'] },
+        { id: '360011519218', value: ticket['support_topics'] },
+        { id: '360011762698', value: ticket['telephone_number'] },
```
Comment (aldavidson) [inline app/services/zendesk_service.rb:26]: I guess these IDs have been created in the Zendesk user interface? I'm a little uncomfortable having them hard-coded within the method body here, could you maybe pull them out into a constant at the top? Something like:

```
class ZendeskService
  attr_accessor :ticket

  CUSTOM_FIELD_IDS = {
    contact_form: '360011490478',
    user_type: '360011798678',
    support_topics: '360011519218',
    telephone_number: '360011762698',
  }.freeze

(...)

  def send!
    ZendeskAPI::Request.create!(
      (...)
      custom_fields: [
        { id: CUSTOM_FIELD_IDS[:contact_form], value: 'contact_form' },
        { id: CUSTOM_FIELD_IDS[:user_type], value: ticket['user_type'] },
        { id: CUSTOM_FIELD_IDS[:support_topics], value: ticket['support_topics'] },
        { id: CUSTOM_FIELD_IDS[:telephone_number], value: ticket['telephone_number'] },
    ],
```

File: spec/form_objects/support_ticket/check_your_request_form_spec.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                                    .with({
+                                      'school_name' => 'School 1',
+                                      'school_unique_id' => '123456',
+                                      'subject' => 'ONLINE FORM - (123456) School 1',
+                                    })
+      end
+    end
+  end
+
+  describe 'when not in production and znedesk creditials are not provided' do
```
Comment (aldavidson) [inline spec/form_objects/support_ticket/check_your_request_form_spec.rb:50]:
```suggestion
  describe 'when zendesk credentials are not provided' do
```
Typo, and removed 'in production' as there's no mention of production in the test code

File: spec/form_objects/support_ticket/check_your_request_form_spec.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      'email_address' => 'joe@bloggs.com',
+      'user_type' => 'parent',
+      'telephone_number' => '0207 333 4444',
+      'subject' => 'My query',
+      'message' => 'This is my query',
+      'support_topics' => %w[hello world],
+    }
+  end
+
+  describe 'when in production and znedesk creditials are provided' do
```
Comment (aldavidson) [inline spec/form_objects/support_ticket/check_your_request_form_spec.rb:16]:
```suggestion
  describe 'when Zendesk credentials are provided' do
```
Typo, and removed 'in production' as there's no mention of production in the test code

**Review by aldavidson — CHANGES_REQUESTED** [review]
Just a few typos and a minor suggestion about moving the hard-coded IDs to a constant, otherwise all good :+1:

**Review by aldavidson — APPROVED** [review]

**Pull request merged**